pkg/platform/kube/platform.go

```diff
@@ -1308,6 +1275,146 @@ func (p *Platform) enrichFunctionsWithAPIGateways(ctx context.Context, functions
+	}
+
+	// no preemption mode was selected, default to prevent
+	if functionConfig.Spec.PreemptionMode == "" {
+		p.Logger.DebugWithCtx(ctx, "No preemption mode was given, defaulting to prevent")
+		functionConfig.Spec.PreemptionMode = functionconfig.RunOnPreemptibleNodesPrevent
+	}
+
+	p.Logger.DebugWithCtx(ctx,
+		"Enriching function spec for given preemption mode",
+		"functionName", functionConfig.Meta.Name,
+		"preemptionMode", functionConfig.Spec.PreemptionMode)
+
+	switch functionConfig.Spec.PreemptionMode {
+	case functionconfig.RunOnPreemptibleNodesPrevent:
+
+		// ensure no preemptible node tolerations
+		functionConfig.PruneTolerations(preemptibleNodes.Tolerations)
+		functionConfig.PruneTolerations(preemptibleNodes.GPUTolerations)
+
+		// if tolerations were given, purge `affinity` preemption related configuration
+		if preemptibleNodes.Tolerations != nil {
+			functionConfig.
+				PruneAffinityNodeSelectorRequirement(
+					preemptibleNodes.CompileAffinityByLabelSelector(v1.NodeSelectorOpIn))
+
+			// prevention is done by tolerations (as spec was explicitly given)
+			// and thus, stop here
+			break
+		}
+
+		// initial affinity
+		if functionConfig.Spec.Affinity == nil {
+			functionConfig.Spec.Affinity = &v1.Affinity{}
+		}
+
+		// initial node affinity
+		if functionConfig.Spec.Affinity.NodeAffinity == nil {
+			functionConfig.Spec.Affinity.NodeAffinity = &v1.NodeAffinity{}
+		}
+
+		// using a single term with potentially multiple expressions to ensure affinity.
+		// when having multiple terms, pod scheduling is succeeded if at least one
+		// term is satisfied.
+		functionConfig.
+			Spec.
+			Affinity.
+			NodeAffinity.
+			RequiredDuringSchedulingIgnoredDuringExecution = &v1.NodeSelector{
+			NodeSelectorTerms: preemptibleNodes.CompileAntiAffinityByLabelSelectorNoScheduleOnMatchingNodes(),
+		}
+
+	case functionconfig.RunOnPreemptibleNodesConstrain:
+
+		// enrich with tolerations
+		functionConfig.EnrichWithTolerations(preemptibleNodes.Tolerations)
+
+		// in case function pod requires gpu resource(s), enrich with gpu tolerations
+		if functionConfig.Spec.PositiveGPUResourceLimit() {
+			functionConfig.EnrichWithTolerations(preemptibleNodes.GPUTolerations)
+		}
+
+		// initial affinity
+		if functionConfig.Spec.Affinity == nil {
+			functionConfig.Spec.Affinity = &v1.Affinity{}
+		}
+
+		// initial node affinity
+		if functionConfig.Spec.Affinity.NodeAffinity == nil {
+			functionConfig.Spec.Affinity.NodeAffinity = &v1.NodeAffinity{}
+		}
+
+		// using a single term with potentially multiple expressions to ensure affinity.
+		// when having multiple terms, pod scheduling is succeeded if at least one
+		// term is satisfied.
+		functionConfig.
+			Spec.
+			Affinity.
+			NodeAffinity.
+			RequiredDuringSchedulingIgnoredDuringExecution = &v1.NodeSelector{
+			NodeSelectorTerms: preemptibleNodes.CompileAffinityByLabelSelectorScheduleOnOneOfMatchingNodes(),
+		}
```

Comment:
You're missing prune anti affinity here no ?

Reply:
well, not really because it overrides the anti affinity by adding affinity on exact same fields. I will add for correctness and readability for tbh it will have 0 impact.

Reply:
I got confused, you can see that for pods affinity and anti affinity are different fields, for nodes apparently it's not like that, there's only node affinity, there's no such thing node anti affinity, it's just an affinity with a not in operator